lib/THC/THCGeneral.c

```diff
@@ -42,21 +101,153 @@ void THCudaInit(THCState* state)
 
             continue;
           }
+
           THCudaCheck(err);
         }
       }
     }
   }
-  THCudaCheck(cudaSetDevice(device));
+
+  /* Restore previous device before continuing */
+  THCudaCheck(cudaSetDevice(prevDev));
 }
 
-void THCudaShutdown(THCState* state)
+int THCState_getNumDevices(THCState *state)
 {
-  THCRandom_shutdown(state);
-  free(state->blasState);
-  free(state->rngState);
-  free(state->deviceProperties);
-  THCudaBlas_shutdown(state);
+  return state->numDevices;
+}
+
+void THCState_reserveStreams(THCState* state, int numStreams)
+{
+  if (numStreams <= state->numUserStreams)
+  {
+    return;
+  }
+
+  int prevDev = -1;
+  THCudaCheck(cudaGetDevice(&prevDev));
+
+  /* Otherwise, we have to allocate a new set of streams */
+  cudaStream_t** newStreams =
+    (cudaStream_t**) malloc(state->numDevices * sizeof(cudaStream_t*));
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    THCudaCheck(cudaSetDevice(dev));
+
+    /* +1 for the default stream as well */
+    newStreams[dev] =
+      (cudaStream_t*) malloc((numStreams + 1) * sizeof(cudaStream_t));
+
+    /* Copy over old streams
+       (0 is default stream, 1 ... numUserStreams are rest) */
+    for (int stream = 0; stream <= state->numUserStreams; ++stream) {
+      newStreams[dev][stream] = state->streamsPerDevice[dev][stream];
+    }
+
+    /* Allocate new streams */
+    for (int stream = state->numUserStreams + 1; stream <= numStreams; ++stream) {
+      newStreams[dev][stream] = NULL;
+      THCudaCheck(cudaStreamCreate(&newStreams[dev][stream]));
+    }
+  }
+
+  cudaStream_t** oldStreams = state->streamsPerDevice;
+  state->streamsPerDevice = newStreams;
+  state->numUserStreams = numStreams;
+
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    free(oldStreams[dev]);
+  }
+  free(oldStreams);
+
+  THCudaCheck(cudaSetDevice(prevDev));
+}
+
+int THCState_getNumStreams(THCState* state)
+{
+  return state->numUserStreams;
+}
+
+void THCState_resetStreams(THCState* state, int device)
+{
+  if (state->currentStream !=
+      state->streamsPerDevice[device][state->currentPerDeviceStream]) {
+    THError("Unexpected stream state");
+  }
+
+  /* Reallocate all streams for the current device; the 0 stream
+     doesn't need updating */
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    for (int stream = 1; stream <= state->numUserStreams; ++stream) {
+      THCudaCheck(cudaStreamCreate(&state->streamsPerDevice[dev][stream]));
```

Comment:
Fair point. I didn't look at it in the context of deviceReset. Maybe we should add a comment to this and other methods saying that they assume that deviceReset has been previously called.

I'd be equally happy to just remove device reset from cutorch.